NightscoutServiceKit/OTPManager.swift

```diff
@@ -0,0 +1,223 @@
+//
+
+import Foundation
+import LoopKit
+import OneTimePassword
+import Base32
+
+private let OTPSecretKeyService = "OTPSecretKeyService"
+private let OTPSecretKeyCreatedService = "OTPSecretKeyCreatedService"
+private let OTPRecentAcceptedPasswordsService = "OTPRecentAcceptedPasswordsService"
+
+public protocol OTPSecretStore {
+    
+    func tokenSecretKey() -> String?
+    func setTokenSecretKey(_ key: String?) throws
+    
+    func tokenSecretKeyName() -> String?
+    func setTokenSecretKeyName(_ name: String?) throws
+    
+    func recentAcceptedPasswords() -> [String]
+    func setRecentAcceptedPasswords(_ passwords: [String]) throws
+}
+
+extension KeychainManager: OTPSecretStore {
+    
+    public func tokenSecretKey() -> String? {
+        return try? getGenericPasswordForService(OTPSecretKeyService)
+    }
+    
+    public func setTokenSecretKey(_ key: String?) throws {
+        try replaceGenericPassword(key, forService: OTPSecretKeyService)
+    }
+    
+    public func tokenSecretKeyName() -> String? {
+        return try? getGenericPasswordForService(OTPSecretKeyCreatedService)
+    }
+    
+    public func setTokenSecretKeyName(_ name: String?) throws {
+        try replaceGenericPassword(name, forService: OTPSecretKeyCreatedService)
+    }
+    
+    public func recentAcceptedPasswords() -> [String] {
+        guard let recentString = try? getGenericPasswordForService(OTPRecentAcceptedPasswordsService) else {
+            return []
+        }
+
+        return convertRecentAcceptedPasswordsFromString(recentString)
+    }
+    
+    public func setRecentAcceptedPasswords(_ passwords: [String]) throws {
+        try replaceGenericPassword(convertRecentAcceptedPasswordsToString(passwords), forService: OTPRecentAcceptedPasswordsService)
+    }
+    
+    func convertRecentAcceptedPasswordsToString(_ recentAcceptedPasswords: [String]) -> String {
+        return recentAcceptedPasswords.joined(separator: ",")
+    }
+    
+    func convertRecentAcceptedPasswordsFromString(_ passwordsString: String) -> [String] {
+        return passwordsString.split(separator: ",").map({String($0)})
+    }
+}
+
+public class OTPManager {
+    
+    private var secretStore: OTPSecretStore
+    private var nowDateSource: () -> Date
+    let algorithm: Generator.Algorithm = .sha512
+    var tokenPeriod: TimeInterval
+    var passwordDigitCount = 6
+    public static var defaultTokenPeriod: TimeInterval = 30
+    
+    public init(secretStore: OTPSecretStore = KeychainManager(), nowDateSource: @escaping () -> Date = {Date()}, tokenPeriod: TimeInterval = OTPManager.defaultTokenPeriod) {
+        self.secretStore = secretStore
+        self.nowDateSource = nowDateSource
+        self.tokenPeriod = tokenPeriod
+        if secretStore.tokenSecretKey() == nil || secretStore.tokenSecretKeyName() == nil {
+            resetSecretKey()
+        }
+    }
+    
+    public func validateOTP(otpToValidate: String) -> Bool {
```

Comment:
We hang onto the last 2 codes as we need to make sure a user can't reuse an OTP code per OTP guidelines.